qcodes/dataset/data_set_protocol.py

```diff
@@ -162,10 +218,286 @@ def export_info(self) -> ExportInfo:
+                path=file_path, engine="h5netcdf", invalid_netcdf=True
+            )
+        else:
+            xarr_dataset.to_netcdf(path=file_path, engine="h5netcdf")
+        return file_path
+
+    def _export_as_csv(self, path: str, file_name: str) -> str:
+        """Export data as csv to a given path with file prefix."""
+        dfdict = self.to_pandas_dataframe_dict()
+        dataframe_to_csv(
+            dfdict=dfdict,
+            path=path,
+            single_file=True,
+            single_file_name=file_name,
+        )
+        return os.path.join(path, file_name)
+
+    @staticmethod
+    def _validate_parameters(
+        *params: Union[str, ParamSpec, _BaseParameter]
+    ) -> List[str]:
+        """
+        Validate that the provided parameters have a name and return those
+        names as a list.
+        The Parameters may be a mix of strings, ParamSpecs or ordinary
+        QCoDeS parameters.
+        """
+
+        valid_param_names = []
+        for maybe_param in params:
+            if isinstance(maybe_param, str):
+                valid_param_names.append(maybe_param)
+            else:
+                try:
+                    maybe_param_name = maybe_param.name
+                except Exception as e:
+                    raise ValueError("This parameter does not have  a name") from e
+                valid_param_names.append(maybe_param_name)
+        return valid_param_names
+
+    @staticmethod
+    def _reshape_array_for_cache(
+        param: ParamSpecBase, param_data: np.ndarray
+    ) -> np.ndarray:
+        """
+        Shape cache data so it matches data read from database.
+        This means:
+
+        - Add an extra singleton dim to array data
+        - flatten non array data into a linear array.
+        """
+        param_data = np.atleast_1d(param_data)
+        if param.type == "array":
+            new_data = np.reshape(param_data, (1,) + param_data.shape)
+        else:
+            new_data = param_data.ravel()
+        return new_data
+
+    def run_timestamp(self, fmt: str = "%Y-%m-%d %H:%M:%S") -> Optional[str]:
+        """
+        Returns run timestamp in a human-readable format
+
+        The run timestamp is the moment when the measurement for this run
+        started. If the run has not yet been started, this function returns
+        None.
+
+        Consult with :func:`time.strftime` for information about the format.
+        """
+        return raw_time_to_str_time(self.run_timestamp_raw, fmt)
+
+    def completed_timestamp(self, fmt: str = "%Y-%m-%d %H:%M:%S") -> Optional[str]:
+        """
+        Returns timestamp when measurement run was completed
+        in a human-readable format
+
+        If the run (or the dataset) is not completed, then returns None.
+
+        Consult with ``time.strftime`` for information about the format.
+        """
+        return raw_time_to_str_time(self.completed_timestamp_raw, fmt)
+
+
+class DataSetTypes(str, Enum):
```

Comment:
shall the name of the class be singular in the same spirit as Color.RED and SmuError.VoltageComplianceHit?
```suggestion
class DataSetType(str, Enum):
```

Reply:
yes probably better.